src/wp-includes/block-template.php

```diff
@@ -228,7 +228,28 @@ function get_the_block_template_html() {
 	$content = $wp_embed->autoembed( $content );
 	$content = shortcode_unautop( $content );
 	$content = do_shortcode( $content );
-	$content = do_blocks( $content );
+
+	/*
+	 * Most block themes omit the `core/query` and `core/post-template` blocks in their singular content templates.
+	 * While this technically still works since singular content templates are always for only one post, it results in
+	 * the main query loop never being entered which causes bugs in core and the plugin ecosystem.
+	 *
+	 * The workaround below ensures that the loop is started even for those singular templates. The while loop will by
+	 * definition only go through a single iteration, i.e. `do_blocks()` is only called once.
+	 *
+	 * Even if the block template contained a `core/query` and `core/post-template` block referencing the main query
+	 * loop, it would not cause errors since it would use a cloned instance and go through the same loop of a single
+	 * post, within the actual main query loop.
+	 */
+	if ( is_singular() ) {
+		while ( have_posts() ) {
```

Comment:
Since we're expecting this to only run once anyway, would it be safer for us to use an `if` here? Also, is there a possibility that a singular template could load in a bizarre scenario where `have_posts()` would return `false` and then no content is returned? I can't think of an edge case where that would occur, but seems technically possible if someone has modified the main query somehow.

Reply:
Ah, I didn't remember that the final `have_posts()` call is what also sets `in_the_loop` to `false` ([among other things](https://github.com/WordPress/wordpress-develop/blob/6.3/src/wp-includes/class-wp-query.php#L3675-L3704)). In that case, leaving it as you currently has it makes more sense. 👍🏻

Reply:
SG regarding the `while`, however I left another comment in https://github.com/WordPress/wordpress-develop/pull/5104#discussion_r1309421673 regarding the other thing mentioned above.